_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,161 @@
+	}
+}
+```
+
+To check whether your contract is actually affected, you need to trigger the condition leading to the assembly ``return(...)`` or ``stop()`` and verify that all storage writes that should have happened in the code leading to it are correctly performed.
+
+   
+## Technical Details
+
+The Yul optimizer step responsible for the bug is meant to remove storage writes that it can determine to be redundant. A storage write is considered redundant in either of the following two cases:
+- A subsequent write unconditionally overwrites the written value.
+- The transaction unconditionally reverts after the write.
+
+So similarly to above, the general pattern is:
+1. a storage write, then
+2. potentially unrelated code, and finally 
+3. either another storage write overwriting the write in (1) or an unconditional revert.
+
+The bug is due to the treatment of certain function calls in (2).
+
+If a function call is performed between (1) and (3), the optimizer has to consider the control flow behaviour of the function call, e.g. whether control flow may continue beyond the call, whether the called function always reverts or always successfully terminates. However, in the case that the control flow conditionally may continue after the call to a function, but the function call may also terminate the transaction (see examples below), the optimizer incorrectly still behaved, as if the control flow *always* continued after the function.
+
+Hence a subsequent overwrite or unconditional revert in (3) results in the write in (1) to be removed, even though (2) may in fact terminate and is supposed to retain the storage write in (1).
+
+So on the Yul level the issue manifests if the following sequence occurs within a Yul block during optimization:
+
+1. A storage write occurs (e.g. ``sstore(0, 1)``).
+2. A user-defined function is called and this function both has a control flow path that terminates the transaction and a control flow path out of the function.
+3. Either of the following:
+  - Another storage write occurs that makes the first storage write above redundant (e.g. ``sstore(0, 2)``).
+  - The control flow unconditionally reverts (e.g. ``revert(0, 0)``)
+
+In this situation, the optimizer will treat the first storage write as redundant, despite the fact that the called Yul function may successfully terminate the EVM call, in which case the first storage write is in fact not redundant at all.
+
+The following is a full example of a affected Yul snippet:
+
+```
+{
+  function f() {
+    if gt(calldatasize(), 4) { leave }
+    return(0, 0)
+  }
+  sstore(0, 1) // This sstore will incorrectly be removed.
+  f()
+  sstore(0, 2)
+}
+```
+
+At the call of ``f()``, the first ``sstore`` should already be marked as required by the optimizer, i.e. it should not be considered for removal regardless of what happens afterwards, since the control flow path of ``f()`` reaching the ``return(0, 0)`` already relies on the ``sstore`` happening.
```

Comment:
How about this? "Should be" is kinda ambiguous to me. Might refer to something that should have happened but didn't or something that did happen :)
```suggestion
In the absence of the bug, the first ``sstore`` would have been marked as required by the optimizer upon encountering the call to ``f()``, i.e. not considered for removal regardless of what happens afterwards, since the control flow path of ``f()`` reaching the ``return(0, 0)`` already relies on the ``sstore`` happening.
```